kvbc/include/v4blockchain/detail/blocks.h

```diff
@@ -0,0 +1,111 @@
+#include "bcstatetransfer/SimpleBCStateTransfer.hpp"
+#include <vector>
+#include "categorized_kvbc_msgs.cmf.hpp"
+#include "assertUtils.hpp"
+#include "categorization/updates.h"
+
+namespace concord::kvbc::v4blockchain::detail {
+/*
+Block is the unit that composes the block chain.
+It recursively links the state by including the digest of the previous block.
+In this implementation, the block will contains all data i.e. the set of key-values.
+its layout is as follows
+|VERSION|DIGEST|KEY-VALUES|
+it uses a buffer, and fast access due to the known locations.
+It is stored as is in DB and is a passable unit for the use of state-transfer.
+*/
+class Block {
+ public:
+  using version = std::array<char, 2>;
+  static constexpr version BLOCK_VERSION = {0xf, 0x1};
+  static constexpr uint64_t HEADER_SIZE = sizeof(version) + sizeof(concord::util::digest::BlockDigest);
+  // Pre-reserve buffer size
+  Block(uint64_t reserve_size) : buffer_(HEADER_SIZE, 0) {
+    buffer_.reserve(reserve_size);
+    addVersion();
+  }
+  Block() : Block(HEADER_SIZE) {}
+
+  Block(const std::string& buffer) : buffer_(buffer.size()) {
+    std::copy(buffer.cbegin(), buffer.cend(), buffer_.begin());
+  }
+
+  Block(const Block&) = default;
+  Block(Block&&) = default;
+  Block& operator=(const Block&) & = default;
+  Block& operator=(Block&&) & = default;
+
+  const version& getVersion() const {
+    ConcordAssert(isValid_);
+    ConcordAssert(buffer_.size() >= sizeof(version));
+    return *reinterpret_cast<const version*>(buffer_.data());
+  }
+
+  const concord::util::digest::BlockDigest& parentDigest() const {
+    ConcordAssert(isValid_);
+    ConcordAssert(buffer_.size() >= HEADER_SIZE);
+    return *reinterpret_cast<const concord::util::digest::BlockDigest*>(buffer_.data() + sizeof(version));
+  }
+
+  void addDigest(const concord::util::digest::BlockDigest& digest) {
+    ConcordAssert(isValid_);
+    ConcordAssert(buffer_.size() >= HEADER_SIZE);
+    std::copy(digest.cbegin(), digest.cend(), buffer_.data() + sizeof(BLOCK_VERSION));
+  }
+
+  concord::util::digest::BlockDigest calculateDigest(concord::kvbc::BlockId id) const {
+    ConcordAssert(isValid_);
+    return calculateDigest(id, reinterpret_cast<const char*>(buffer_.data()), buffer_.size());
+  }
+
+  void addUpdates(const concord::kvbc::categorization::Updates& category_updates);
+
+  concord::kvbc::categorization::Updates getUpdates() const;
+
+  const std::vector<uint8_t>& getBuffer() const {
+    ConcordAssert(isValid_);
+    return buffer_;
+  }
+
+  std::vector<uint8_t>&& moveBuffer() {
+    ConcordAssert(isValid_);
+    isValid_ = false;
+    return std::move(buffer_);
+  }
+
+  static concord::util::digest::BlockDigest calculateDigest(concord::kvbc::BlockId id,
+                                                            const char* buffer,
+                                                            uint64_t size) {
+    return bftEngine::bcst::computeBlockDigest(id, buffer, size);
+  }
+
```

Comment:
Should there be a function called validateBlock as below:
//Assume that the block is |{digest}|{version}|{block data:list of key-value pairs}
bool validateBlock() {
  //Assume that isValid_ is false if the ctor with buffer called and block is not validated
  isValid_ = <compute the block digest and then compare that with the stored digest>
  return isValid_;
}